k-es-grpc/src/main/kotlin/no/ks/kes/grpc/GrpcEventSubscriberFactory.kt

```diff
@@ -0,0 +1,112 @@
+package no.ks.kes.grpc
+
+import com.eventstore.dbclient.*
+import io.grpc.StatusRuntimeException
+import mu.KotlinLogging
+import no.ks.kes.grpc.GrpcEventUtil.isIgnorable
+import no.ks.kes.grpc.GrpcEventUtil.isResolved
+import no.ks.kes.grpc.GrpcSubscriptionDroppedReason.*
+import no.ks.kes.lib.*
+import no.ks.kes.lib.EventData
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicLong
+import kotlin.reflect.KClass
+
+private val log = KotlinLogging.logger {}
+
+class GrpcEventSubscriberFactory(
+    private val eventStoreDBClient: EventStoreDBClient,
+    private val serdes: EventSerdes,
+    private val category: String,
+    private val metadataSerdes: EventMetadataSerdes<out Metadata>? = null
+) : EventSubscriberFactory<GrpcSubscriptionWrapper> {
+
+    override fun getSerializationId(eventDataClass: KClass<EventData<*>>): String =
+        serdes.getSerializationId(eventDataClass)
+
+    override fun createSubscriber(
+        subscriber: String,
+        fromEvent: Long,
+        onEvent: (EventWrapper<EventData<*>>) -> Unit,
+        onClose: (Exception) -> Unit,
+        onLive: () -> Unit
+    ): GrpcSubscriptionWrapper {
+
+        val streamId = "\$ce-$category"
+        //val revision = StreamRevision(fromEvent)
+        // TODO: No way to start subscription from StreamRevision.END without knowing the number of events (same as for Esjc based client)
+        val revision = when {
+            fromEvent == -1L -> StreamRevision.START
+            fromEvent > -1L -> StreamRevision(fromEvent)
+            else -> error("the from-event $fromEvent is invalid, must be a number equal to or larger than -1")
+        }
+
+        val listener = object : SubscriptionListener() {
+
+            var lastEventProcessed = AtomicLong(-1)
+
+            override fun onEvent(subscription: Subscription, resolvedEvent: ResolvedEvent) {
+                //if (event.getEvent().position == Position.END)
+                // TODO: No way to say if we're live and should call onLive
```

Comment:
TODO, hva blir løsningen her?

Reply:
Usikker på om dette er noe som faktisk brukes av noen k-es applikasjoner, men om vi trenger det kan vi legge inn en lesing av slutten på streamen først, for å se hvilket eventnummer som er siste når subscription opprettes, og så kalle onLive når vi er kommet til den eventen. Gjør noe tilsvarende i eventstore2, for der blir onLive faktisk brukt.